### ISSUE TITLE
How to use `gather_for_metrics()` with decoder-generated strings to compute rouge score?

### ISSUE THREAD

**Issue by plamb-viso** [huggingface/accelerate#2064]
I am fine-tuning an encoder-decoder model and during the validation step, using the `.generate` method to generate tokens from the decoder that are subsequently decoded into strings (in this case classes). These generations are occurring across 8 GPUs and I am using Accelerate to manage the distribution.

My hope was to append these strings to lists, and pass the lists to `gather_for_metrics()` on each GPU to get a "master list" of predictions and references, added to the rouge metric and then computed:

```python
predictions, references = accelerator.gather_for_metrics(
            (predictions, references)
        )

rouge_metric.add_batch(
    predictions=predictions,
    references=references,
)

rouge_score = rouge_metric.compute(rouge_types=["rougeL"], use_aggregator=True)["rougeL"]
```

After encountering some strange errors, i noticed that `gather_for_metrics()` will [only interact with tensors](https://huggingface.co/docs/accelerate/v0.19.0/en/package_reference/accelerator#accelerate.Accelerator.gather_for_metrics)

And from what I can tell, you cannot create a torch.Tensor with string members.

How do the accelerate folks recommend using `gather_for_metrics()` with decoder-generated strings?

**Comment by BenjaminBossan** [comment]
Would it be possible for you to collect the model outputs (I guess logits) using `gather_for_metrics` and do the decoding into tokens afterwards?

**Comment by muellerzr** [comment]
Iirc we are also looking at adding support for `gather_object` out of torch distributed (it may already be in main)

**Comment by muellerzr** [comment]
@plamb-viso try installing from main via `pip install git+https://github.com/huggingface/accelerate`, we should already support this

**Comment by plamb-viso** [comment]
@muellerzr this suggestion appears to have gotten me past the errors; however the way I was using `gather_for_metrics()` previously (for non-decoder models) does not appear to work the same. For example (using a decoder model):

```python
On GPU 1:
predictions = ["prediction1", "prediction2"]
references = ["reference1", "reference2"]

On GPU 2:
predictions = ["prediction3", "prediction4"]
references = ["reference3", "reference4"]

predictions, references = accelerator.gather_for_metrics(
    (predictions, references)
)
```
Results in:
```python
[["prediction1", "prediction2"], ["reference1", "reference2"], ["prediction3", "prediction4"], ["reference3", "reference4"]]
```
My expectation was that it would be:
```python
[["prediction1", "prediction2", "prediction3", "prediction4"], ["reference1", "reference2", "reference3", "reference4"]]
```
Am I passing the data to `gather_for_metrics()` incorrectly? I could iterate the result and create what I expect, but is that the intention of this function?

**Comment by muellerzr** [comment]
That *seems* to be weird, @plamb-viso can I assume that your strings here are fillers for the tensors? Or are they actual strings. `gather_object` may work slightly differently than `gather`
